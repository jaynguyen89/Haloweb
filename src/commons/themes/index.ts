import { ThemeNight } from './night';
import { ThemeDay } from './day';
import { ThemeDream } from './dream';
import { ThemeEarth } from './earth';
import { ThemeGrass } from './grass';
import { ThemeOcean } from './ocean';
import { ThemeRose } from './rose';
import { ThemeSky } from './sky';
import { ThemeSunflower } from './sunflower';
import { ThemeTangerine } from './tangerine';

export default {
    ThemeNight,
    ThemeDay,
    ThemeOcean,
    ThemeDream,
    ThemeSky,
    ThemeGrass,
    ThemeSunflower,
    ThemeTangerine,
    ThemeRose,
    ThemeEarth,
};