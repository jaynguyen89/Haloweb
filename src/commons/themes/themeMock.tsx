import React from 'react';
import { ThemeProvider } from '@mui/material';
import { createTheme } from '@mui/material/styles';
import { render } from '@testing-library/react';
import { shallow, mount } from 'enzyme';
import colors from 'src/commons/colors';

export const paletteMock = {
    text: {
        primary: colors.lightBlack,
        secondary: colors.darkGrey,
        disabled: colors.grey,
    },
    primary: {
        main: colors.lightGrey,
        light: colors.lighterGrey,
        dark: colors.darkGrey,
        contrastText: colors.darkerGrey,
    },
    secondary: {
        main: colors.darkTeal,
        light: colors.lightTeal,
        dark: colors.deepTeal,
        contrastText: colors.lighterTeal,
    },
    error: {
        main: colors.darkRed,
        light: colors.lightRed,
        dark: colors.deepRed,
        contrastText: colors.lighterRed,
    },
    warning: {
        main: colors.darkYellow,
        light: colors.lightYellow,
        dark: colors.deepYellow,
        contrastText: colors.lighterYellow,
    },
    info: {
        main: colors.darkBlue,
        light: colors.lightBlue,
        dark: colors.deepBlue,
        contrastText: colors.lighterBlue,
    },
    success: {
        main: colors.darkGreen,
        light: colors.lightGreen,
        dark: colors.deepGreen,
        contrastText: colors.lighterGreen,
    },
    common: {
        white: colors.white,
        black: colors.black,
    },
};

export const MockTheme = ({ children }: { children: React.ReactElement }) => {
    const themeMock = createTheme({
        mixins: {
            toolbar: { width: '100%' },
            shadowLight: '0 3px 6px rgba(255,255,255,0.16), 0 3px 6px rgba(255,255,255,0.23)',
            shadowDark: '0 3px 6px rgba(0,0,0,0.16), 0 3px 6px rgba(0,0,0,0.23)',
            shadowLightUp: '0 -3px 6px rgba(255,255,255,0.2)',
            shadowDarkUp: '0px -3px 6px rgba(0, 0, 0, 0.2)',
        },
        breakpoints: {
            values: {
                xs: 0,
                sm: 667,
                md: 992,
                lg: 1280,
                xl: 1536,
            },
        },
        palette: paletteMock,
    });

    return <ThemeProvider theme={themeMock}>{children}</ThemeProvider>;
}

export const renderWithTheme = (component: React.ReactElement) => render(
    <MockTheme>{component}</MockTheme>
);

export const shallowWithTheme = (component: React.ReactElement) => shallow(
    <MockTheme>{component}</MockTheme>
);

export const mountWithTheme = (component: React.ReactElement) => mount(
    <MockTheme>{component}</MockTheme>
);
