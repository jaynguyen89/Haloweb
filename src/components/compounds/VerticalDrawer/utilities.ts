import { IconName } from '@fortawesome/free-regular-svg-icons';
import { IconDefinition } from '@fortawesome/free-solid-svg-icons';

export type TVerticalDrawerMenuItem = {
    isActive?: true,
    text: string,
    icon?: IconName | IconDefinition | string,
    link?: string,
    onClick?: () => void,
    subMenu?: Array<Omit<TVerticalDrawerMenuItem, 'subMenu'>>,
};

export type TVerticalDrawerMenu = {
    title: string,
    items: Array<TVerticalDrawerMenuItem>,
    divider?: true,
};

export const setActiveMenuItem = (
    menuItems: Array<TVerticalDrawerMenu>,
    keyword: string,
    searchBy: 'text' | 'link',
    inSubMenu: boolean = false,
): Array<TVerticalDrawerMenu> => {
    /* eslint-disable max-depth */
    for (const menuItem of menuItems)
        for (const item of menuItem.items)
            if (item.isActive) {
                item.isActive = undefined;

                if (item.hasOwnProperty('subMenu'))
                    for (const subMenuItem of item.subMenu!)
                        if (subMenuItem.isActive) subMenuItem.isActive = undefined;

                break;
            }
    /* eslint-enable max-depth */

    if (!inSubMenu) {
        for (const menuItem of menuItems)
            for (const item of menuItem.items) {
                if (searchBy === 'text' && item.text.includes(keyword))
                    item.isActive = true;

                if (searchBy === 'link' && item.link!.includes(keyword))
                    item.isActive = true;

                if (item.isActive) break;
            }

        return menuItems;
    }

    for (const menuItem of menuItems)
        for (const item of menuItem.items)
            if (item.hasOwnProperty('subMenu')) {
                for (const subMenuItem of item.subMenu!) {
                    if (searchBy === 'text' && subMenuItem.text.includes(keyword))
                        subMenuItem.isActive = true;

                    if (searchBy === 'link' && item.link!.includes(keyword))
                        subMenuItem.isActive = true;

                    if (subMenuItem.isActive) {
                        item.isActive = true;
                        break;
                    }
                }
            }

    return menuItems;
};
