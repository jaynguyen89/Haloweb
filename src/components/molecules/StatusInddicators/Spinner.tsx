import React from 'react';
import { Box, CircularProgress, CircularProgressProps, Typography } from '@mui/material';
import { useIsStageIncluded } from 'src/hooks/useStage';

type TSpinner = CircularProgressProps & {
    stage: string,
};

const Spinner = ({
    stage,
    color,
    size = 100,
    thickness = 2,
}: TSpinner) => {
    const visible = useIsStageIncluded(stage);
    if (!visible) return null;

    return (
        <CircularProgress
            variant='indeterminate'
            color={color}
            size={size}
            thickness={thickness}
        />
    );
};

export default Spinner;

type TLabelSpinner = CircularProgressProps & {
    stage: string,
    progress: number,
    textVariant?: string,
};

export const LabelSpinner = ({
    stage,
    progress,
    color,
    size = 100,
    thickness = 2,
    textVariant = 'h6',
}: TLabelSpinner) => {
    const visible = useIsStageIncluded(stage);
    if (!visible) return null;

    return (
        <Box sx={{ position: 'relative', display: 'inline-flex' }}>
            <CircularProgress
                variant='determinate'
                color={color}
                size={size}
                thickness={thickness}
                value={progress}
            />
            
            <Box
                sx={{
                    top: 0,
                    left: 0,
                    bottom: 0,
                    right: 0,
                    position: 'absolute',
                    display: 'flex',
                    alignItems: 'center',
                    justifyContent: 'center',
                }}
            >
                <Typography
                    /* eslint-disable  @typescript-eslint/no-explicit-any */
                    variant={textVariant as any}
                    component='div'
                    color='text.secondary'
                >{progress}%</Typography>
            </Box>
        </Box>
    );
};
