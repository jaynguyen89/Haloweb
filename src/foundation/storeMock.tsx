import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import React from 'react';
import { publicDataMock } from 'src/models/PublicData';

export const stageMock = {
    stages: [
        { name: 'LOADING_TEST' },
        {
            name: 'SHOW_FLASHER_SERVER_ERROR',
            type: 'error',
            message: 'This is server error test.',
        },
        {
            name: 'STAGE_FLASHER_TEST',
            message: '{{what}} Test',
            messageParams: { what: 'StageFlasher' },
            type: 'success',
        },
    ],
};

const mockStore = configureStore([]);
const storeMock = mockStore({
    languageStore: {
        siteLanguage: {
            selected: 'en',
            detected: 'en',
        },
    },
    publicDataStore: {
        publicData: publicDataMock,
    },
    stageStore: stageMock,
});

export const withStoreMock = (component: React.ReactElement) => {
    return (
        <Provider store={storeMock}>
            {component}
        </Provider>
    );
};
