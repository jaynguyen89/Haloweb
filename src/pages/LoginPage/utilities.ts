import { IAuthenticationData } from 'src/models/Authentication';
import { IRegionalizedPhoneNumber } from 'src/models/Profile';
import IPublicData from 'src/models/PublicData';
import {
    TFormDataState, TRangeOption, TSpecialOption,
    TValidatorOption,
    TValidatorOptionsMapFn,
    ValidatorNames,
} from 'src/utilities/data-validators/dataValidators';

export enum LoginFormFieldNames {
    EmailAddress = 'EmailAddress',
    AreaCode = 'AreaCode',
    PhoneNumber = 'PhoneNumber',
    Password = 'Password',
    Trusted = 'Trusted',
}

export const initialLoginFormDataState: TFormDataState<typeof LoginFormFieldNames> = {
    [LoginFormFieldNames.EmailAddress]: {
        value: undefined,
        caption: undefined,
    },
    [LoginFormFieldNames.AreaCode]: {
        value: undefined,
    },
    [LoginFormFieldNames.PhoneNumber]: {
        value: undefined,
        caption: undefined,
    },
    [LoginFormFieldNames.Password]: {
        value: undefined,
        caption: undefined,
    },
    [LoginFormFieldNames.Trusted]: { value: false },
};

export const loginFieldValidatorMap = {
    [LoginFormFieldNames.EmailAddress]: ValidatorNames.RangeValidator,
    [LoginFormFieldNames.AreaCode]: ValidatorNames.RangeValidator,
    [LoginFormFieldNames.PhoneNumber]: ValidatorNames.RangeValidator,
    [LoginFormFieldNames.Password]: ValidatorNames.RangeValidator,
};

export type TLoginFieldKey = Omit<typeof LoginFormFieldNames, 'Trusted'>;

export const loginValidatorOptionsMapFn: TValidatorOptionsMapFn<TLoginFieldKey> =
    (publicData?: IPublicData): TValidatorOption<TLoginFieldKey> =>
({
    [LoginFormFieldNames.EmailAddress]: {
        min: 6,
        max: 100,
    } as TRangeOption,
    [LoginFormFieldNames.AreaCode]: {
        among: publicData!.countries.map(country => country.telephoneCode),
    } as TRangeOption,
    [LoginFormFieldNames.PhoneNumber]: {
        equal: 9,
        numbersOnly: true,
    } as TRangeOption,
    [LoginFormFieldNames.Password]: {
        min: 8,
        max: 24,
    } as TRangeOption,
});

export const createLoginData = (formData: TFormDataState<typeof LoginFormFieldNames>): IAuthenticationData => ({
    emailAddress: formData[LoginFormFieldNames.EmailAddress].value as string | undefined,
    phoneNumber: formData[LoginFormFieldNames.PhoneNumber].value === undefined ? undefined : {
        phoneNumber: formData[LoginFormFieldNames.PhoneNumber].value,
        regionCode: formData[LoginFormFieldNames.AreaCode].value,
    } as IRegionalizedPhoneNumber,
    password: formData[LoginFormFieldNames.Password].value,
    isTrusted: formData[LoginFormFieldNames.Trusted].value,
    deviceInformation: undefined,
});
