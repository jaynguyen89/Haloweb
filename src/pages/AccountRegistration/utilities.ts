import { TRangeOption } from 'src/utilities/dataValidators';

export enum RegistrationValidatorFieldNames {
    EmailAddress = 'EmailAddress',
    AreaCode = 'AreaCode',
    PhoneNumber = 'PhoneNumber',
    Password = 'Password',
    PasswordConfirm = 'PasswordConfirm',
    Username = 'Username',
    Gender = 'Gender',
    GivenName = 'GivenName',
    MiddleName = 'MiddleName',
    FamilyName = 'FamilyName',
    FullName = 'FullName',
}

export const initialRegistrationFormDataState = {
    [RegistrationValidatorFieldNames.EmailAddress]: {
        value: undefined,
        caption: undefined,
    },
    [RegistrationValidatorFieldNames.AreaCode]: {
        value: undefined,
    },
    [RegistrationValidatorFieldNames.PhoneNumber]: {
        value: undefined,
        caption: undefined,
    },
    [RegistrationValidatorFieldNames.Password]: {
        value: undefined,
        caption: undefined,
    },
    [RegistrationValidatorFieldNames.PasswordConfirm]: {
        value: undefined,
    },
    [RegistrationValidatorFieldNames.Username]: {
        value: undefined,
        caption: undefined,
    },
    [RegistrationValidatorFieldNames.Gender]: {
        value: 4,
        caption: undefined,
    },
    [RegistrationValidatorFieldNames.GivenName]: {
        value: undefined,
        caption: undefined,
    },
    [RegistrationValidatorFieldNames.MiddleName]: {
        value: undefined,
        caption: undefined,
    },
    [RegistrationValidatorFieldNames.FamilyName]: {
        value: undefined,
        caption: undefined,
    },
    [RegistrationValidatorFieldNames.FullName]: {
        value: undefined,
        caption: undefined,
    },
};

export const validatorOptions = {
    [RegistrationValidatorFieldNames.EmailAddress]: {

    },
    [RegistrationValidatorFieldNames.AreaCode]: {

    },
    [RegistrationValidatorFieldNames.PhoneNumber]: {

    },
    [RegistrationValidatorFieldNames.Password]: {

    },
    [RegistrationValidatorFieldNames.Username]: {

    },
    [RegistrationValidatorFieldNames.Gender]: {

    },
    [RegistrationValidatorFieldNames.GivenName]: {

    },
    [RegistrationValidatorFieldNames.MiddleName]: {

    },
    [RegistrationValidatorFieldNames.FamilyName]: {

    },
    [RegistrationValidatorFieldNames.FullName]: {

    },
};
