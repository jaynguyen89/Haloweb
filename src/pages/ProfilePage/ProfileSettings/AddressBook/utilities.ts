import { IEasternAddress, IWesternAddress } from 'src/models/Address';
import { AddressVariant } from 'src/models/enums/apiEnums';
import {
    TFormDataState,
    TValidatorOption,
    TValidatorOptionsMapFn,
    ValidatorNames,
} from 'src/utilities/data-validators/dataValidators';
import IPublicData from 'src/models/PublicData';
import { toPascalCase } from 'src/utilities/stringUtilities';
import _cloneDeep from 'lodash/cloneDeep';

export enum WesternAddressFormFields {
    Variant = 'Variant',
    BuildingName = 'BuildingName',
    PoBoxNumber = 'PoBoxNumber',
    StreetAddress = 'StreetAddress',
    Suburb = 'Suburb',
    Postcode = 'Postcode',
    DivisionId = 'DivisionId',
    CountryId = 'CountryId',
}

export enum EasternAddressFormFields {
    Variant = 'Variant',
    BuildingName = 'BuildingName',
    PoBoxNumber = 'PoBoxNumber',
    StreetAddress = 'StreetAddress',
    Lane = 'Lane',
    Group = 'Group',
    Quarter = 'Quarter',
    Hamlet = 'Hamlet',
    Commute = 'Commute',
    Ward = 'Ward',
    District = 'District',
    Town = 'Town',
    City = 'City',
    DivisionId = 'DivisionId',
    CountryId = 'CountryId',
}

export const initialWesternAddressFormState: TFormDataState<typeof WesternAddressFormFields> = {
    [WesternAddressFormFields.Variant]: { value: AddressVariant.Western },
    [WesternAddressFormFields.BuildingName]: {
        value: undefined,
        caption: undefined,
    },
    [WesternAddressFormFields.PoBoxNumber]: {
        value: undefined,
        caption: undefined,
    },
    [WesternAddressFormFields.StreetAddress]: {
        value: undefined,
        caption: undefined,
    },
    [WesternAddressFormFields.Suburb]: {
        value: undefined,
        caption: undefined,
    },
    [WesternAddressFormFields.Postcode]: {
        value: undefined,
        caption: undefined,
    },
    [WesternAddressFormFields.DivisionId]: {
        value: undefined,
        caption: undefined,
    },
    [WesternAddressFormFields.CountryId]: {
        value: undefined,
        caption: undefined,
    },
};

export const initialEasternAddressFormState: TFormDataState<typeof EasternAddressFormFields> = {
    [EasternAddressFormFields.Variant]: { value: AddressVariant.Eastern },
    [EasternAddressFormFields.BuildingName]: {
        value: undefined,
        caption: undefined,
    },
    [EasternAddressFormFields.PoBoxNumber]: {
        value: undefined,
        caption: undefined,
    },
    [EasternAddressFormFields.StreetAddress]: {
        value: undefined,
        caption: undefined,
    },
    [EasternAddressFormFields.Lane]: {
        value: undefined,
        caption: undefined,
    },
    [EasternAddressFormFields.Group]: {
        value: undefined,
        caption: undefined,
    },
    [EasternAddressFormFields.Quarter]: {
        value: undefined,
        caption: undefined,
    },
    [EasternAddressFormFields.Hamlet]: {
        value: undefined,
        caption: undefined,
    },
    [EasternAddressFormFields.Commute]: {
        value: undefined,
        caption: undefined,
    },
    [EasternAddressFormFields.Ward]: {
        value: undefined,
        caption: undefined,
    },
    [EasternAddressFormFields.District]: {
        value: undefined,
        caption: undefined,
    },
    [EasternAddressFormFields.Town]: {
        value: undefined,
        caption: undefined,
    },
    [EasternAddressFormFields.City]: {
        value: undefined,
        caption: undefined,
    },
    [EasternAddressFormFields.DivisionId]: {
        value: undefined,
        caption: undefined,
    },
    [EasternAddressFormFields.CountryId]: {
        value: undefined,
        caption: undefined,
    },
};

export const getDefaultWesternAddressData = (address: IWesternAddress): TFormDataState<typeof WesternAddressFormFields> => {
    const westernAddressData = _cloneDeep(initialWesternAddressFormState);
    Object.keys(address).forEach(key => {
        if (key !== 'variant') westernAddressData[toPascalCase(key)] = {
            value: address[key],
        };
    });
    
    return westernAddressData;
};

export const getDefaultEasternAddressData = (address: IEasternAddress): TFormDataState<typeof EasternAddressFormFields> => {
    const easternAddressData = _cloneDeep(initialEasternAddressFormState);
    Object.keys(address).forEach(key => {
        if (key !== 'variant') easternAddressData[toPascalCase(key)] = {
            value: address[key],
        };
    });

    return easternAddressData;
};

export const westernAddressFormValidatorMap = {
    [WesternAddressFormFields.BuildingName]: ValidatorNames.RangeValidator,
    [WesternAddressFormFields.PoBoxNumber]: ValidatorNames.RangeValidator,
    [WesternAddressFormFields.StreetAddress]: ValidatorNames.RangeValidator,
    [WesternAddressFormFields.Suburb]: ValidatorNames.RangeValidator,
    [WesternAddressFormFields.Postcode]: ValidatorNames.RangeValidator,
    [WesternAddressFormFields.DivisionId]: ValidatorNames.RangeValidator,
    [WesternAddressFormFields.CountryId]: ValidatorNames.RangeValidator,
};

export const easternAddressFormValidatorMap = {
    [EasternAddressFormFields.BuildingName]: ValidatorNames.RangeValidator,
    [EasternAddressFormFields.PoBoxNumber]: ValidatorNames.RangeValidator,
    [EasternAddressFormFields.StreetAddress]: ValidatorNames.RangeValidator,
    [EasternAddressFormFields.Lane]: ValidatorNames.RangeValidator,
    [EasternAddressFormFields.Group]: ValidatorNames.RangeValidator,
    [EasternAddressFormFields.Quarter]: ValidatorNames.RangeValidator,
    [EasternAddressFormFields.Hamlet]: ValidatorNames.RangeValidator,
    [EasternAddressFormFields.Commute]: ValidatorNames.RangeValidator,
    [EasternAddressFormFields.Ward]: ValidatorNames.RangeValidator,
    [EasternAddressFormFields.District]: ValidatorNames.RangeValidator,
    [EasternAddressFormFields.Town]: ValidatorNames.RangeValidator,
    [EasternAddressFormFields.City]: ValidatorNames.RangeValidator,
    [EasternAddressFormFields.DivisionId]: ValidatorNames.RangeValidator,
    [EasternAddressFormFields.CountryId]: ValidatorNames.RangeValidator,
};

export const westernAddressFormValidatorOptionsMapFn: TValidatorOptionsMapFn<TWesternAddressFormKeys> =
    (publicData?: IPublicData, extra?: unknown): TValidatorOption<TWesternAddressFormKeys> =>
({

});

export const easternAddressFormValidatorOptionsMapFn: TValidatorOptionsMapFn<TEasternAddressFormKeys> =
    (publicData?: IPublicData, extra?: unknown): TValidatorOption<TEasternAddressFormKeys> =>
({

});