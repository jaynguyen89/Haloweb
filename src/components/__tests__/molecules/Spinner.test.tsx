import React from 'react';
import { configure, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from '@cfaester/enzyme-adapter-react-18';
import '@testing-library/jest-dom';
import { withStoreMock } from 'src/foundation/storeMock';
import Spinner, { LabelSpinner, TLabelSpinner, TSpinner } from 'src/components/molecules/StatusIndicators/Spinner';
import { renderWithTheme } from 'src/commons/themes/themeMock';
import { toPascalCase } from 'src/utilities/stringUtilities';
import { queryByRole } from '@testing-library/react';

configure({ adapter: new Adapter() });

describe('Spinner.tsx', () => {
    it('renders successfully', () => {
        const wrapper = shallow(withStoreMock(<Spinner stage='showcase' />));
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should not be visible when stage is incorrect', () => {
        const wrapper = renderWithTheme(withStoreMock(<Spinner stage='invisible' />));
        const progressElement = wrapper.container.getElementsByClassName('MuiCircularProgress-root')[0];
        expect(progressElement).toBeUndefined();
    });

    it('renders component in various colors', () => {
        const propses: Array<TSpinner> = [
            {
                color: 'primary',
                size: 80,
                thickness: 2,
            },
            {
                color: 'secondary',
                size: 100,
                thickness: 2,
            },
            {
                color: 'error',
                size: 120,
                thickness: 3,
            },
            {
                color: 'info',
                size: 140,
                thickness: 3,
            },
            {
                color: 'success',
                size: 160,
                thickness: 4,
            },
            {
                color: 'warning',
                size: 180,
                thickness: 4,
            },
        ];

        const wrappers = propses.map(props => renderWithTheme(withStoreMock(<Spinner stage='showcase' {...props} />)));

        propses.forEach((props, i) => {
            const progressBar = queryByRole(wrappers[i].container, 'progressbar');
            expect(progressBar.classList.value.includes('MuiCircularProgress-root')).toBeTruthy();
            expect(progressBar.classList.value.includes('MuiCircularProgress-indeterminate')).toBeTruthy();
            expect(progressBar.classList.value.includes(`MuiCircularProgress-color${toPascalCase(props.color)}`)).toBeTruthy();
            expect(progressBar.style.width).toBe(`${props.size}px`);
            expect(progressBar.style.height).toBe(`${props.size}px`);

            const svg = progressBar?.getElementsByTagName('svg')[0];
            expect(svg).toBeDefined();

            const circle = svg.getElementsByTagName('circle')[0];
            expect(circle).toBeDefined();
            expect(circle.classList.value.includes('MuiCircularProgress-circle')).toBeTruthy();
            expect(circle.classList.value.includes('MuiCircularProgress-circleIndeterminate')).toBeTruthy();
            expect(circle.attributes.getNamedItem('stroke-width').value).toEqual(`${props.thickness}`);
        });
    });
});

describe('LabelSpinner.tsx', () => {
    it('renders successfully', () => {
        const wrapper = shallow(withStoreMock(<LabelSpinner stage='showcase' progress={50} />));
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should not be visible when stage is incorrect', () => {
        const wrapper = renderWithTheme(withStoreMock(<Spinner stage='invisible' progress={50} />));
        const progressElement = wrapper.container.getElementsByClassName('MuiBox-root')[0];
        expect(progressElement).toBeUndefined();
    });

    it('renders component in various colors', () => {
        const propses: Array<TLabelSpinner> = [
            {
                color: 'primary',
                size: 80,
                thickness: 2,
                progress: 25,
            },
            {
                color: 'secondary',
                size: 100,
                thickness: 2,
                progress: 35,
            },
            {
                color: 'error',
                size: 120,
                thickness: 3,
                progress: 45,
            },
            {
                color: 'info',
                size: 140,
                thickness: 3,
                progress: 55,
            },
            {
                color: 'success',
                size: 160,
                thickness: 4,
                progress: 65,
            },
            {
                color: 'warning',
                size: 180,
                thickness: 4,
                progress: 75,
            },
        ];

        const wrappers = propses.map(props => renderWithTheme(withStoreMock(<LabelSpinner stage='showcase' {...props} />)));

        propses.forEach((props, i) => {
            const progressBar = queryByRole(wrappers[i].container, 'progressbar');
            expect(progressBar.classList.value.includes('MuiCircularProgress-root')).toBeTruthy();
            expect(progressBar.classList.value.includes('MuiCircularProgress-determinate')).toBeTruthy();
            expect(progressBar.classList.value.includes(`MuiCircularProgress-color${toPascalCase(props.color)}`)).toBeTruthy();
            expect(progressBar.style.width).toBe(`${props.size}px`);
            expect(progressBar.style.height).toBe(`${props.size}px`);
            expect(progressBar.attributes.getNamedItem('aria-valuenow').value).toEqual(`${props.progress}`);

            const svg = progressBar.getElementsByTagName('svg')[0];
            expect(svg).toBeDefined();

            const circle = svg.getElementsByTagName('circle')[0];
            expect(circle).toBeDefined();
            expect(circle.classList.value.includes('MuiCircularProgress-circle')).toBeTruthy();
            expect(circle.classList.value.includes('MuiCircularProgress-circleDeterminate')).toBeTruthy();
            expect(circle.attributes.getNamedItem('stroke-width').value).toEqual(`${props.thickness}`);

            const progressLabel = wrappers[i].container.getElementsByClassName('MuiTypography-root')[0];
            expect(progressLabel).toBeDefined();
            expect(progressLabel.classList.value.includes('MuiTypography-h6')).toBeTruthy();
            expect(progressLabel.textContent).toEqual(`${props.progress}%`);
        });
    });
});
