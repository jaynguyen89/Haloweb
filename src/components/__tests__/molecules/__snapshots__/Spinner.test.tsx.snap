// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LabelSpinner renders successfully 1`] = `
<ContextProvider
  value={
    Object {
      "getServerState": undefined,
      "noopCheck": "once",
      "stabilityCheck": "once",
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Object {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <LabelSpinner
    progress={50}
    stage="showcase"
  />
</ContextProvider>
`;

exports[`LabelSpinner.tsx renders successfully 1`] = `
<ContextProvider
  value={
    Object {
      "getServerState": undefined,
      "noopCheck": "once",
      "stabilityCheck": "once",
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Object {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <LabelSpinner
    progress={50}
    stage="showcase"
  />
</ContextProvider>
`;

exports[`Spinner renders successfully 1`] = `
<ContextProvider
  value={
    Object {
      "getServerState": undefined,
      "noopCheck": "once",
      "stabilityCheck": "once",
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Object {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <Spinner
    stage="showcase"
  />
</ContextProvider>
`;

exports[`Spinner.tsx renders successfully 1`] = `
<ContextProvider
  value={
    Object {
      "getServerState": undefined,
      "noopCheck": "once",
      "stabilityCheck": "once",
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Object {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <Spinner
    stage="showcase"
  />
</ContextProvider>
`;
